### DISCUSSION TITLE
Is backref=backref() necessary when defining relationship?

### DISCUSSION THREAD

**Discussion by yishang11** [sqlalchemy/sqlalchemy#7889] [Usage Questions]
Hi,
I am using sqlite with sqlalchemy. 
In the online document 
(https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html)
it seems we need to do 
`address = relationship("Address",backref=backref("user", uselist=False))` if we want user and address to have 1 to 1 relationship.
But when I use above code, uselist has no effect, and user.address can still be a list of 2.
I instead tried `address = relationship("Address", backref="user", uselist=False)`, which worked as expected: when user is associated with 2 addresses, user.address gives the latest one only, and a warning is issued.

My question is: do we really need `backref=backref()` syntax when we want to add argument like uselist, cascade, etc? To me using them without `backref()` worked just fine.
Thank you!

**Comment by zzzeek** [comment]
backrefs are optional.   it's also clearer to use back_populates.  In the document linked at https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#one-to-many:

1. the first example does not use any backref or back_populates.   this illustrates a basic one-to-many
2. the second example illustrates how to combine both one-to-many **and** many-to-one at the same time.  if this pattern is desired, one would usually want to tell the ORM these two relationships are pointing to the same thing in the database, and this is achieved by using the ``back_populates`` keyword.
3. there is then another short example, illustrating that `backref()` is a popular "shorthand" convention for `back_populates`.     Again, completely optional, and if configuration is confusing here, I would advise using only `back_populates` which is intended to make configuration of other options like "uselist" and "cascades" to be clearer.

if you are looking to use "one-to-one", the second example at https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#one-to-one illustrates this most clearly:

```py
class Parent(Base):
    __tablename__ = 'parent'
    id = Column(Integer, primary_key=True)

    # previously one-to-many Parent.children is now
    # one-to-one Parent.child
    child = relationship("Child", back_populates="parent", uselist=False)

class Child(Base):
    __tablename__ = 'child'
    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer, ForeignKey('parent.id'))

    # many-to-one side remains, see tip below
    parent = relationship("Parent", back_populates="child")
```

**Reply by yishang11** [reply]
Thank you so much for your clarification! 
I do like backref because I only need to specify relationship on 1 side. 
Could you confirm if my usage `address = relationship("Address", backref="user", uselist=False)` is correct? Or do I have to put uselist inside backref(): `address = relationship("Address",backref=backref("user", uselist=False))`, which does not work for me at the moment? 
Regarding online resources on the usage of backref inside relationship(), I have seen both `backref=backref()` and bare `backref = $string`, like  (https://medium.com/dataexplorations/sqlalchemy-orm-a-more-pythonic-way-of-interacting-with-your-database-935b57fd2d4d).

**Reply by zzzeek** [reply]
> Thank you so much for your clarification! I do like backref because I only need to specify relationship on 1 side. Could you confirm if my usage `address = relationship("Address", backref="user", uselist=False)` is correct? 

that is correct.  relationship("Address") naturally refers to one-to-many, so by setting uselist=False, we flip the default of uselist=True.

> Regarding online resources on the usage of backref inside relationship(), I have seen both `backref=backref()` and bare `backref = $string`, like (https://medium.com/dataexplorations/sqlalchemy-orm-a-more-pythonic-way-of-interacting-with-your-database-935b57fd2d4d).

backref() is vastly more popular since it's been around for decades and is also less typing.   but for reasons like the confusion you have here, I try to encourage people to use back_populates and only use backref() if they understand what it means.
